File: frontend/src/api/productPriceBreakdown.ts
/**
 * Product Price Breakdown V1 — read-model over CPP + EIC (no second calculator).
 */

import { getAPIBaseURL } from "@/lib/config";
import { formatApiErrorResponse } from "@/lib/apiError";

export type PriceBreakdownLineGroup =
  | "material"
  | "machine"
  | "labor"
  | "service"
  | "ai_decision"
  | "adjustment"
  | "commercial"
  | "internal";

export interface PriceBreakdownLine {
  line_id: string;
  line_group: PriceBreakdownLineGroup;
  resource_code: string;
  display_name: string;
  quantity_key?: string | null;
  formula_display: string;
  quantity?: number | null;
  unit?: string | null;
  base_value?: number | null;
  currency?: string | null;
  source_type: string;
  source_id?: string | null;
  minimum?: number | null;
  waste?: number | null;
  adjustment?: number | null;
  internal_cost?: number | null;
  commercial_value?: number | null;
  cpp_line?: string | null;
  eic_rule?: string | null;
  configurable: boolean;
  warning?: string | null;
  confidence?: string | null;
  rationale_ro?: string | null;
  ai_decision_id?: string | null;
}

export interface PriceBreakdownGroupTotal {
  line_group: PriceBreakdownLineGroup;
  line_count: number;
  internal_subtotal?: number | null;
  commercial_subtotal?: number | null;
  currency?: string | null;
}

export interface PriceBreakdownTotals {
  material_internal?: number | null;
  machine_internal?: number | null;
  labor_internal?: number | null;
  service_internal?: number | null;
  consumables_internal?: number | null;
  overhead_internal?: number | null;
  ai_contribution_note_ro?: string | null;
  internal_total?: number | null;
  commercial_subtotal?: number | null;
  commercial_total?: number | null;
  currency: string;
  cpp_total_matches: boolean;
  eic_total_matches: boolean;
  no_duplicate_commercial_codes: boolean;
  no_duplicate_internal_codes: boolean;
}

export interface ProductPriceBreakdownResponse {
  schema_version: string;
  template_code: string;
  configuration_id: string;
  fixture_id?: string | null;
  currency: string;
  ownership_note_ro: string;
  publication_status?: string | null;
  operational_readiness?: string | null;
  uses_ai_defaults: boolean;
  configuration_summary: Record<string, unknown>;
  lines: PriceBreakdownLine[];
  group_totals: PriceBreakdownGroupTotal[];
  totals: PriceBreakdownTotals;
  ai_decisions: Array<Record<string, unknown>>;
  calibration_hooks: Array<{
    line_code?: string | null;
    estimated_minutes?: number | null;
    purpose?: string | null;
    excluded_from_total: boolean;
    note_ro: string;
  }>;
  cpp_status?: string | null;
  eic_status?: string | null;
  warnings: string[];
  blockers: string[];
  eic_provenance: Array<Record<string, unknown>>;
  cpp_provenance: Array<Record<string, unknown>>;
  acm_treatments_blocked?: boolean | null;
}

export const productPriceBreakdownApi = {
  postBreakdown: async (
    templateCode: string,
    body?: {
      workspace_id?: string | null;
      quote_input?: Record<string, unknown> | null;
      currency?: string;
      fixture_id?: string | null;
    },
  ): Promise<ProductPriceBreakdownResponse> => {
    const res = await fetch(
      `${getAPIBaseURL()}/api/v1/product-system/templates/${encodeURIComponent(templateCode)}/price-breakdown`,
      {
        method: "POST",
        credentials: "include",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body ?? {}),
      },
    );
    if (!res.ok) {
      const detail = await formatApiErrorResponse(
        res,
        `price breakdown failed: HTTP ${res.status}`,
      );
      throw new Error(detail);
    }
    return res.json();
  },
};

File: backend/schemas/product_price_breakdown.py
"""PRODUCT_PRICE_BREAKDOWN_V1 — read-model over CPP + EIC + recipe (no second calculator)."""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

PRODUCT_PRICE_BREAKDOWN_VERSION = "1.0.0"

LineGroup = Literal[
    "material",
    "machine",
    "labor",
    "service",
    "ai_decision",
    "adjustment",
    "commercial",
    "internal",
]

SourceType = Literal[
    "market",
    "catalog",
    "owner",
    "AI_DECISION",
    "measured",
    "legacy",
    "documented_commercial",
    "inventory",
    "unknown",
]


class PriceBreakdownLine(BaseModel):
    line_id: str
    line_group: LineGroup
    resource_code: str
    display_name: str
    quantity_key: Optional[str] = None
    formula_display: str
    quantity: Optional[float] = None
    unit: Optional[str] = None
    base_value: Optional[float] = None
    currency: Optional[str] = None
    source_type: SourceType = "unknown"
    source_id: Optional[str] = None
    minimum: Optional[float] = None
    waste: Optional[float] = None
    adjustment: Optional[float] = None
    internal_cost: Optional[float] = None
    commercial_value: Optional[float] = None
    cpp_line: Optional[str] = None
    eic_rule: Optional[str] = None
    configurable: bool = False
    warning: Optional[str] = None
    confidence: Optional[str] = None
    rationale_ro: Optional[str] = None
    ai_decision_id: Optional[str] = None
    # MATERIAL_MARKET_PRICE_REGISTRY_V1 — additive purchase provenance
    material_source_type: Optional[str] = None
    material_supplier: Optional[str] = None
    material_freshness: Optional[str] = None
    material_effective_from: Optional[str] = None
    material_normalization_formula: Optional[str] = None
    material_normalized_unit: Optional[str] = None
    material_normalized_price: Optional[float] = None
    material_canonical: Optional[bool] = None


class PriceBreakdownGroupTotal(BaseModel):
    line_group: LineGroup
    line_count: int = 0
    internal_subtotal: Optional[float] = None
    commercial_subtotal: Optional[float] = None
    currency: Optional[str] = None


class PriceBreakdownTotals(BaseModel):
    material_internal: Optional[float] = None
    machine_internal: Optional[float] = None
    labor_internal: Optional[float] = None
    service_internal: Optional[float] = None
    consumables_internal: Optional[float] = None
    overhead_internal: Optional[float] = None
    ai_contribution_note_ro: Optional[str] = None
    internal_total: Optional[float] = None
    commercial_subtotal: Optional[float] = None
    commercial_total: Optional[float] = None
    currency: str = "RON"
    cpp_total_matches: bool = False
    eic_total_matches: bool = False
    no_duplicate_commercial_codes: bool = True
    no_duplicate_internal_codes: bool = True


class PriceBreakdownCalibrationHook(BaseModel):
    """Secondary time/calibration metadata — not required for totals."""

    line_code: Optional[str] = None
    estimated_minutes: Optional[float] = None
    purpose: Optional[str] = None
    excluded_from_total: bool = True
    note_ro: str = "Timpul este evidență secundară pentru calibrare, nu baza de cost."


class ProductPriceBreakdownResponse(BaseModel):
    schema_version: str = PRODUCT_PRICE_BREAKDOWN_VERSION
    template_code: str
    configuration_id: str
    fixture_id: Optional[str] = None
    currency: str = "RON"
    ownership_note_ro: str = (
        "Linia de referință a laboratorului: Cost producție (EIC). "
        "Preț comercial (CPP) rămâne vizibil pentru reconciliere, nu ca ofertă. "
        "Desfășurătorul nu recalculează. Materialele folosesc cost achiziție. "
        "Manopera folosește driveri fizici. Timpul e secundar. "
        "Fără adaos / ofertă / comandă în finish line."
    )
    publication_status: Optional[str] = None
    operational_readiness: Optional[str] = None
    uses_ai_defaults: bool = False
    configuration_summary: dict[str, Any] = Field(default_factory=dict)
    lines: list[PriceBreakdownLine] = Field(default_factory=list)
    group_totals: list[PriceBreakdownGroupTotal] = Field(default_factory=list)
    totals: PriceBreakdownTotals = Field(default_factory=PriceBreakdownTotals)
    ai_decisions: list[dict[str, Any]] = Field(default_factory=list)
    calibration_hooks: list[PriceBreakdownCalibrationHook] = Field(default_factory=list)
    cpp_status: Optional[str] = None
    eic_status: Optional[str] = None
    warnings: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(default_factory=list)
    eic_provenance: list[dict[str, Any]] = Field(default_factory=list)
    cpp_provenance: list[dict[str, Any]] = Field(default_factory=list)
    acm_treatments_blocked: Optional[bool] = None
